Add new BatteryMonitor widget

andkem opened the pull request:
Adds a widget called BatteryMonitor for monitoring the status of the
battery in a laptop. It is written with the aim to provide a structure
that facilitates porting to other systems, such as Linux.

Currently it only supports FreeBSD.

I wrote this for the reasons stated in https://github.com/qtile/qtile/issues/1274. This one only works on FreeBSD because I currently don't have a Linux laptop, but I am willing to work around the issue if no one else feels the call to help port Linux support.

If Linux support is ported to this one, I guess one could rename it and simply replace the old Battery widget.

flacjacket commented on libqtile/widget/battery_monitor.py:
```diff
@@ -0,0 +1,394 @@
+from .. import images
+
+
+class State(Enum):
```
I'd `@unique` decorate this.

andkem replied:
Done

flacjacket commented on libqtile/widget/battery_monitor.py:
```diff
@@ -0,0 +1,394 @@
+
+
+class State(Enum):
+    CHARGING = 0
```
Might be good to make all of these `auto()`

andkem replied:
Done

flacjacket commented on libqtile/widget/battery_monitor.py:
```diff
@@ -0,0 +1,394 @@
+    UNKNOWN = 4
+
+
+class _Battery(object):
```
I actually like this a lot more as an interface that we could port the other battery over to. It would be awesome if you could also add some type annotations to it!

It'd be good to make this an `abc.ABCMeta` abstract class and all of the not implemented methods as abstract methods.

andkem replied:
Done

andkem replied:
I wrote the data type in the comment. If you want them in some specific format, please give me an example.

flacjacket commented on libqtile/widget/__init__.py:
```diff
@@ -38,6 +38,7 @@ def safe_import(module_name, class_name):
 
 safe_import("backlight", "Backlight")
 safe_import("battery", ["Battery", "BatteryIcon"])
+safe_import("battery_monitor", ["BatteryMonitor", "BatteryMonitorIcon"])
```
Why put this in a different file? Can we just adjust the battery to add this functionality via your FreeBSD implementation?

andkem replied:
I put it in a separate file because it was simpler while working on it. If or when the Linux integration happens, I was planning on putting it back in the old file. I believe I explained it some in the PR or the issue that spawned the PR.

flacjacket commented on libqtile/widget/battery_monitor.py:
```diff
@@ -0,0 +1,394 @@
+
+        stat = stat.group(1)
+        if stat == 'charging':
+            result = State.CHARGING
```
No need to assign a variable in all of these, just return the values

andkem replied:
I'm of the opinion that assigning a value to a result and returning it at the end of a function is cleaner and more maintainable as it reduces code complexity. If you feel strongly about it, I can of course change it.

flacjacket replied:
No problem, I'll leave that up to preference.

andkem commented:
I've now rewritten the Linux parts as well as I finally found my USB stick and was able to install Linux on it. I'll get to addressing all the comments and taking a few clean-up sweeps next.

flacjacket commented:
Awesome, thanks for this! Let us know when you are happy with it and I'll give it a shot!

andkem commented:
> Awesome, thanks for this! Let us know when you are happy with it and I'll give it a shot!

I'm happy with the code as it is now. I only have a computer that reports current on Linux via sysfs. The part that handles power is thus untested. If someone has a computer reporting power, it would be great if they could try it out.

ramnes commented:
It works: ![itworks.png](https://i.imgur.com/yKx2U4d.png)

ramnes commented:
Although, there is one thing I miss with this new version of the widget, which is that it used to write `Full` rather than `= 99% 0:00 0.00 W`.

flacjacket commented:
Check out #1293 and let me know what you think. I started from this PR and built off of that, and maintained exactly the behavior ramnes mentioned, with a config to change it back.

andkem commented:
> Check out #1293 and let me know what you think. I started from this PR and built off of that, and maintained exactly the behavior ramnes mentioned, with a config to change it back.

I think it looks good. I appreciate making "Full" an option as I prefer to see the text. I've run it on my FreeBSD machine and it still seems to work fine with your changes.

Feel free to refactor or rewrite as you see fit to get the code up to your standard. You can push changes directly to this PR or close it and replace it with your own. I don't mind either way.

flacjacket commented:
Thanks for this! I opted to just pull in this PR and rebase my tweaks on top of master.

ramnes commented:
Great, thanks @flacjacket!
